File: scripts/kind-books.mjs
// Every kind, at every version, has a BOOK: kinds/<ext>/v<N>.playbook — a playbook, in the version-2
// form, that explains how the kind works at that version. `studio-check --book <ext> [N]` prints its
// path; `studio-check --books` lists them all.
//
//   node scripts/kind-books.mjs           write the book of every kind and version that has none yet —
//                                         the engine's own account (the spec) under an always-on event,
//                                         and the fresh document under another; never overwrites a book
//   node scripts/kind-books.mjs --check   every book exists and passes the checker, and every kind the page
//                                         offers has its field table, kinds/<ext>/v<N>.fields.yaml (exit 1 otherwise) — CI
//
// The checker must be built first: pnpm cli:build.
import { execFileSync } from "node:child_process";
import { existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import path from "node:path";
import { createRequire } from "node:module";

const root = path.resolve(path.dirname(new URL(import.meta.url).pathname.replace(/^\/([A-Za-z]:)/, "$1")), "..");
const cli = path.join(root, "apps", "cli", "dist", "check.cjs");
if (!existsSync(cli)) { console.error("build the checker first: pnpm cli:build"); process.exit(2); }
const yaml = createRequire(path.join(root, "packages", "filekinds", "package.json"))("js-yaml");
// The checker exits 1 when something is missing or failing; its output is still the answer.
const run = (...args) => {
  try { return execFileSync("node", [cli, ...args], { encoding: "utf8", stdio: ["ignore", "pipe", "pipe"] }); }
  catch (e) { if (e.stdout !== undefined && e.status === 1) { if (args[0] === "--books") return e.stdout; throw Object.assign(new Error(e.stdout || e.stderr), { stdout: e.stdout }); } throw e; }
};

const check = process.argv.includes("--check");
const books = JSON.parse(run("--books", "--json"));

let written = 0;
for (const b of books) {
  if (b.exists) continue;
  if (check) continue;
  const spec = run("--spec", b.ext).trim();
  const label = spec.match(/^# \.\w+ — (.+)$/m)?.[1] ?? b.ext;
  const doc = {
    version: 2,
    title: `How a version-${b.version} ${label.toLowerCase()} works`,
    description: `The .${b.ext} kind at version ${b.version}: the engine's own account, and the document a fresh file starts from.`,
    events: [
      {
        key: "always", label: "Always", trigger: "imposed", domain: "Always",
        detail: "What holds at every moment — the engine's account of the kind, the same text studio-check --spec prints.",
        content: { key: "spec", label: "The engine's account", kind: "md", doc: spec + "\n" },
      },
    ],
  };
  mkdirSync(path.dirname(b.path), { recursive: true });
  writeFileSync(b.path, yaml.dump(doc, { lineWidth: -1, noRefs: true }));
  written++;
  console.log(`wrote ${path.relative(root, b.path)}`);
}

// Every book present, and every one a document the checker passes.
const after = JSON.parse(run("--books", "--json"));
const missing = after.filter((b) => !b.exists);
for (const m of missing) console.error(`missing: ${path.relative(root, m.path)}`);
let failed = 0;
for (const b of after.filter((b) => b.exists)) {
  try { run(b.path); } catch (e) { failed++; console.error(String(e.stdout ?? e.message)); }
}

// The FIELD TABLES — kinds/<ext>/v<N>.fields.yaml, the schema as a table — for every kind the page
// offers, at every version: present, YAML, a `fields` list whose entries each name a `field`.
const OFFERED = ["brief", "playbook", "kanban", "calendar", "policy", "flow", "jsonl", "middleware", "pipeline", "collection", "clip", "song", "md"];
let badFields = 0;
for (const b of after.filter((b) => OFFERED.includes(b.ext))) {
  const rel = path.relative(root, b.fields);
  if (!b.fieldsExist) { badFields++; console.error(`missing field table: ${rel}`); continue; }
  try {
    const doc = yaml.load(readFileSync(b.fields, "utf8"));
    const list = doc?.fields;
    if (!Array.isArray(list) || !list.length) throw new Error("no `fields` list");
    for (const f of list) if (!f || typeof f.field !== "string" || !f.field) throw new Error("an entry without a `field`");
    if (doc.kind !== b.ext || doc.version !== b.version) throw new Error(`says kind ${doc.kind} v${doc.version}, sits under ${b.ext} v${b.version}`);
  } catch (e) { badFields++; console.error(`${rel}: ${e.message}`); }
}
console.log(`${after.length} books: ${written} written, ${missing.length} missing, ${failed} failing; field tables: ${badFields} missing or wrong`);
process.exit(missing.length || failed || badFields ? 1 : 0);

File: scripts/kind_books.py
"""Every authoring kind, at every version, has a BOOK (kinds/<ext>/v<N>.playbook — a playbook, in the
version-2 form, that explains how the kind works), a FIELD TABLE (v<N>.fields.yaml) and a JSON SCHEMA
(v<N>.schema.json); the Python package ships copies of all three under studio_kinds/data.

  python scripts/kind_books.py           write the book of every kind and version that has none yet —
                                         the spec under an always-on event; never overwrites a book
  python scripts/kind_books.py --check   every book exists and passes the checker; every field table is
                                         present and well-formed; every schema is present; and the package's
                                         copies of books, field tables, schemas and templates are the
                                         repository's (exit 1 otherwise) — CI
  python scripts/kind_books.py --sync    copy the repository's books, field tables and schemas into the package

Needs the package importable: `pip install -e python` (or run from the repository with PYTHONPATH=python).
"""
from __future__ import annotations

import filecmp
import os
import re
import shutil
import sys

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
sys.path.insert(0, os.path.join(ROOT, "python"))

from studio_kinds import AUTHORING, _yaml, check_path, latest_version, spec  # noqa: E402

KINDS_DIR = os.path.join(ROOT, "kinds")
DATA = os.path.join(ROOT, "python", "studio_kinds", "data")


def repo_paths(ext: str, v: int) -> dict[str, str]:
    return {
        "book": os.path.join(KINDS_DIR, ext, f"v{v}.playbook"),
        "fields": os.path.join(KINDS_DIR, ext, f"v{v}.fields.yaml"),
        "schema": os.path.join(KINDS_DIR, ext, f"v{v}.schema.json"),
    }


def package_paths(ext: str, v: int) -> dict[str, str]:
    return {
        "book": os.path.join(DATA, "books", ext, f"v{v}.playbook"),
        "fields": os.path.join(DATA, "fields", ext, f"v{v}.fields.yaml"),
        "schema": os.path.join(DATA, "schemas", ext, f"v{v}.schema.json"),
    }


def rel(p: str) -> str:
    return os.path.relpath(p, ROOT).replace(os.sep, "/")


def write_missing() -> int:
    written = 0
    for ext in AUTHORING:
        for v in range(1, latest_version(ext) + 1):
            book = repo_paths(ext, v)["book"]
            if os.path.exists(book):
                continue
            text = spec(ext).strip()
            m = re.search(r"^# \.\w+ — (.+)$", text, re.M)
            label = m.group(1) if m else ext
            doc = {
                "version": 2,
                "title": f"How a version-{v} {label.lower()} works",
                "description": f"The .{ext} kind at version {v}: the engine's own account, and the document a fresh file starts from.",
                "events": [{
                    "key": "always", "label": "Always", "trigger": "imposed", "domain": "Always",
                    "detail": "What holds at every moment — the engine's account of the kind, the same text studio-check --spec prints.",
                    "content": {"key": "spec", "label": "The engine's account", "kind": "md", "doc": text + "\n"},
                }],
            }
            os.makedirs(os.path.dirname(book), exist_ok=True)
            with open(book, "w", encoding="utf-8", newline="\n") as f:
                f.write(_yaml.dump(doc))
            written += 1
            print(f"wrote {rel(book)}")
    return written


def sync() -> None:
    for ext in AUTHORING:
        for v in range(1, latest_version(ext) + 1):
            src, dst = repo_paths(ext, v), package_paths(ext, v)
            for k in src:
                if os.path.exists(src[k]):
                    os.makedirs(os.path.dirname(dst[k]), exist_ok=True)
                    shutil.copyfile(src[k], dst[k])
    print("package data synced from kinds/")


def check() -> int:
    problems = 0
    for ext in AUTHORING:
        for v in range(1, latest_version(ext) + 1):
            src, dst = repo_paths(ext, v), package_paths(ext, v)
            if not os.path.exists(src["book"]):
                problems += 1
                print(f"missing: {rel(src['book'])}")
            else:
                r = check_path(src["book"])
                if not r.ok:
                    problems += 1
                    print(f"{rel(src['book'])}: {len(r.problems)} problems")
                    for p in r.problems:
                        print(f"      {p}")
            if not os.path.exists(src["fields"]):
                problems += 1
                print(f"missing field table: {rel(src['fields'])}")
            else:
                try:
                    with open(src["fields"], encoding="utf-8") as f:
                        doc = _yaml.load(f.read())
                    fields = doc.get("fields") if isinstance(doc, dict) else None
                    if not isinstance(fields, list) or not fields:
                        raise ValueError("no `fields` list")
                    for entry in fields:
                        if not isinstance(entry, dict) or not isinstance(entry.get("field"), str) or not entry["field"]:
                            raise ValueError("an entry without a `field`")
                    if doc.get("kind") != ext or doc.get("version") != v:
                        raise ValueError(f"says kind {doc.get('kind')} v{doc.get('version')}, sits under {ext} v{v}")
                except Exception as e:  # noqa: BLE001
                    problems += 1
                    print(f"{rel(src['fields'])}: {e}")
            if not os.path.exists(src["schema"]):
                problems += 1
                print(f"missing schema: {rel(src['schema'])}")
            for k in src:
                if os.path.exists(src[k]) and not (os.path.exists(dst[k]) and filecmp.cmp(src[k], dst[k], shallow=False)):
                    problems += 1
                    print(f"package copy out of step: {rel(dst[k])} — run `python scripts/kind_books.py --sync`")
        tmpl = os.path.join(DATA, "templates", f"untitled.{ext}")
        if not os.path.exists(tmpl):
            problems += 1
            print(f"missing template: {rel(tmpl)}")
    print(f"{problems} problems" if problems else "books, field tables, schemas and package copies all in step")
    return 1 if problems else 0


if __name__ == "__main__":
    if "--sync" in sys.argv:
        sync()
        sys.exit(0)
    if "--check" not in sys.argv:
        write_missing()
    sys.exit(check())
